### PR TITLE
Add atomic gssize

### PR DESCRIPTION
**Pull request by lbudai**
Motivation: sometimes we need a 64 bit counter on 64 bit systems.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by furiel** [comment]
We have both atomic and racy versions for get and set, but the atomic versions are not used by this patchset, only the racy ones. A few questions regarding to this:
- Is there any planned usecase for the atomic versions?
- What is the reason we choose the racy over the atomic versions?
- Being the atomic versions dead code now, should we cover the racy versions instead of the atomic versions in the unit test?

**Comment by lbudai** [comment]
@furiel : discuss it with @MrAnno  :)

I don't really like that we have non-atomic accessors for an atomic type.
I tend to agree that we need to remove all the racy operations, but my focus was not on this, but to change [flow-control mechanism](#2161 ) (and fix a [bug](#2167) ).
I also thought that initiate a new atomic counter will open questions, this is why this is a separate PR :)

My intention with unit test was to cover the _atomic_ operations.... even if there are the non-atomic operations (which I kept to being compatible with existing code). I'm open to merge patches that removes the non-atomic operations from the code, or, that are unit testing the racy operations, but I still want to focus on my other branches.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
Okay, I'm just leaving my comment from #2167 here:
I would like to remove all racy atomic operations from our codebase.
We had only problems with them in the past and I don't see any benefits keeping them.

- g_atomic_counter_racy_get
- atomic_gssize_racy_get
- atomic_gssize_racy_get_unsigned
- atomic_gssize_racy_set

`g_atomic_counter_set` is also secretly racy, I've fixed this in #2045, but I'll open a separate PR for it.

**Comment by MrAnno** [comment]
@lbudai I agree, so I'm approving this PR.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/tests/test_atomic_gssize.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "syslog-ng.h"
+#include "atomic-gssize.h"
```
[the remark this answers is not in the document]
Comment (lbudai) [reply lib/tests/test_atomic_gssize.c]: Between the branches / rebases a change has been lost. I think the correct way is to include `syslog-ng.h`in the `atomic-gssize.h`. I wouldn’t add empty c files.

[the remark this answers is not in the document]
Comment (Kokan) [reply lib/tests/test_atomic_gssize.c]: As I said I would also prefer that the test checks out if header is compilable.
Yes, the `atomic-gssize.h` should and already does include the `syslog-ng.h`, that is fine.
A solution would be to remove the `syslog-ng.h` include from the test.

[the remark this answers is not in the document]
Comment (lbudai) [reply lib/tests/test_atomic_gssize.c]: still not clear what is your change request. simply removing the syslog-ng header from the test won’t solve the problem with the *first* patch. 
changing the ordering is not make any sense if we include one header...

should i meld/squash the include statement into the very first patch?

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/compat/glib.h [hunk 2]
```diff
@@ -53,6 +53,27 @@ GList *g_list_copy_deep (GList *list, GCopyFunc func, gpointer user_data);
...
     (gssize) __sync_fetch_and_add ((atomic), (val));                         \
   }))
+#define g_atomic_pointer_or(atomic, val) \
+  (G_GNUC_EXTENSION ({                                                       \
+    G_STATIC_ASSERT (sizeof *(atomic) == sizeof (gpointer));                 \
+    (void) (0 ? (gpointer) *(atomic) : 0);                                   \
+    (void) (0 ? (val) ^ (val) : 0);                                          \
+    (gsize) __sync_fetch_and_or ((atomic), (val));                           \
+  }))
+#define g_atomic_pointer_xor(atomic, val) \
```
Comment (furiel) [inline lib/compat/glib.h]: All these defines are put `#ifndef g_atomic_pointer_add` guardian.
I miss the `#ifndef g_atomic_pointer_xor` etc. Is it safe to depend everything on `g_atomic_pointer_add`?

Comment (Kokan) [reply lib/compat/glib.h]: At least they have been introduced in the same commit https://github.com/GNOME/glib/commit/83821352657a9481dbff6ab04e8ae60566c17d5e 
But if you question that, I think the check itself also questionable (see if the `g_atomic_pointer_add` has been defined as macro...)

Comment (furiel) [reply lib/compat/glib.h]: Being in the same commit is convincing enough for me. Thanks.

**Pull request merged**